### ISSUE TITLE
Dynamic content block based on URL parameters

### ISSUE THREAD

**Issue by makurj** [uiowa/uiowa#6431]
# Story

As a marketing/CRM web editor, I would like to be able to display a different H1 (headline test) and the header hero image based on the URL parameters.

From Lydia: "We have a campaign launching that has two types of ads / assets:
 
Asset Bundle A: “Future proof your resume!”
Asset Bundle B: “Best in Business”
 
We have urls or QR codes on every ad and billboard, and it’s important that if they scan a QR on a billboard from Bundle A, they see a landing page with the message “Future proof your resume” at the top, and imagery that matches what they saw on the billboard.
 
Therefore we would be required to create duplicates of every asset – two landing pages, two forms, yada yada… When the ONLY thing that differs across the two landing pages is the H1 text and hero image.
 
What the dynamic content by URL params allows us to do is to SIMPLIFY and reduce the effort required to launch a campaign like this (and if we can finish this by mid-June we could even use this feature for the Des Moines campaign that unearthed this user story). It also reduces duplication of efforts and allows us to reuse assets for multiple campaigns and reduces integrations between Drupal and the CRM."

Example (made up) URLs: Note _header_ and _img_ params
https://tippie.uiowa.edu/dsm-landing/?header=Future-Proof%20Your%20Resume&header_img=futureproof&utm_source=cpc&utm_medium=cbj&utm_campaign=tippie23

https://tippie.uiowa.edu/dsm-landing/?header=Future-Proof%20Your%20Resume&utm_source=cpc&utm_medium=cbj&utm_campaign=tippie23 

# Proposed solution

https://www.drupal.org/project/condition_query

**Comment by makurj** [comment]
Tippie’s Des Moines campaign goes live June 26. However, this is not a blocker for the campaign going live. Also, after the feature is live they will need time to get the content set up.

**Comment by quamsta** [comment]
I'd be careful with this. It has the potential for abuse. e.g., someone could look at the URL params, change them, and thus change the page to contain undesirable text. 

Ex: `https://tippie.uiowa.edu/dsm-landing/?header=Bad%20%Text%Here`

**Comment by cory-skeers** [comment]
> I'd be careful with this. It has the potential for abuse. e.g., someone could look at the URL params, change them, and thus change the page to contain undesirable text.
> 
> Ex: `https://tippie.uiowa.edu/dsm-landing/?header=Bad%20%Text%Here`

Good point. My take on it would be to pass the params through some sort of check and have a fallback default. It'd take more setup (params would need to be created and updated both on the campaign side and the website side), but I think it'd be worth the extra effort. And likely want a good default in place anyway, so that the pages can function well coming from a campaign or not.

**Comment by joewhitsitt** [comment]
Is there a user story for this on what is gained by being able to do this?

I mentioned this in zoom chat earlier, but another contrib module (https://www.drupal.org/project/smart_content) could be leveraged or referenced for this issue to do something similar and/or allow us to do a/b testing/personalization in the future.

**Comment by makurj** [comment]
Updated issue summary ^^

Also, this issue has been prioritized over all the other Tippie ones, and completed by early or mid-June.

**Comment by joewhitsitt** [comment]
Been thinking about this a bit. To do anything server-side we would need to wire these up to some kind of cache busting code so that it could fetch something different from the db. particularly if a different media asset is supposed to referenced and rendered instead. To me, that seems dangerous because if discovered, it would open us to uncached hits.

What seems more logical that there is JS that loads with everything it needs before it is cached so it can do some DOM manipulation to existing content for that particular URL request. Having this work with the lazy content however could be hard to time to prevent flickering changes of content. Even so, we would need to capture these conditionals somewhere. (path, parameter trigger, what to change, replacement value, etc.)

I get that this would greatly simplify things for the site editor and reduce their replicated content, but it would greatly complicate our side of things just for a different banner image and text.

**Comment by briand44** [comment]
Grooming conversation... probe into https://www.drupal.org/project/layout_builder_context to see if this could be a potential path forward for this issue. Probe would be a medium.

**Comment by joewhitsitt** [comment]
^^ This looks promising LB Context Integration, but can’t be used on our “header” section and Layout preview would be stacked and not representative though if things had different visibility conditions.

If it works, this would address my concerns about uncached request issues and where this information is stored.

**Comment by briand44** [comment]
Added blocked label based on probe https://github.com/uiowa/uiowa/issues/6473

**Comment by cory-skeers** [comment]
Collating some thoughts from probe #6473 

The primary path forward the team is looking at at this time involves

1. A way to include one or more components on a page which can dynamically be shown or hidden
2. A way to provide context to these components through query parameters

I believe this approach would work well for the use case here, as well as potential future use cases for reusable blocks A/B testing options, and potentially others.

The main blocker on this is with point 1. For point 2, there are several contrib module options which may be suitable and, were those to not fit our needs, if the first part is accounted for, creating a homegrown solution for part 2 is likely a small lift.

The main two options considered for the first point were core functionality that will be introduced as some of the work going into Drupal 10, as well the Layout Builder Context module. Full details below, but for this use case, Layout Builder Context looks like a no-go, whereas future core functionality looks very promising. While we did not create a full proof-of-concept of this type of functionality, I think it gives the right building blocks for us to be able to put it together.

## Where that leaves us:
We are somewhat at the mercy of Drupal core updates and release schedules, as well as the platform's readiness to adopt D10. We would be looking at the timeline of getting block visibility settings from core, and then following up with part 2 of the necessary components to tie visibility to query parameters. This would be the most efficient approach, as there would be little homegrown functionality needed here, though the timeline is unclear.

The other option would be to build a homegrown version of what will eventually be core visibility settings in order to move forward today, with the plan of replacing it with core functionality once that has been adopted. We would have the D10 implementation to base it on, which helps, but there may be dependencies to work around and general refactoring there. Additionally, any homegrown solution should be using the third-party settings on components for the visibility settings, which wouldn't be able to be a one-to-one replacement with core in the future and may add other roadblocks. If we went this route, we'd likely need to do another probe to fully understand the scope of building it, even if the general process is understood.

<details>

<summary> Comparison of upcoming Core visibility and current Layout Builder Context </summary>

# Drupal core block visibility settings

- Tried out Drupal 10 dev with https://www.drupal.org/project/drupal/issues/2916876 with D10 patch (https://www.drupal.org/files/issues/2023-03-10/2916876-167.10x.patch) https://www.drupal.org/project/drupal/issues/2916876#comment-14962025 on simplytest.me.  
  - Works great and seems like better UX than Layout Builder Context because everything is set within the block. 
    - Separate edit tray makes settings these values better UX
    - Unclear if query parameters can be set here, I'm guessing not. 
      - Maybe once this is released there will be more integration? 
    - Request path functionality here seems intended for something like:
      - Only show an apply now button on areas of study pages based on taxonomy terms or url paths. 
    - **No caching issues.**  

# Layout Builder Context
- Tried out https://www.drupal.org/project/layout_builder_context
  - Similar functionality to https://www.drupal.org/project/drupal/issues/2916876
  -  With this module, you have to set a context on the context module page `/admin/structure/context` and **then** set it on each block which might not be the most user-friendly option for webmasters. 
  - Settings are lumped into the block edit screen tray and can make this tray overcrowded.
  - The `request_parameter_condition` module works with `layout_builder_context`, but only worked after I cleared the site cache each time. 
    - https://www.drupal.org/project/request_parameter_condition - supported
      - These modules do the same but don't look very active:
        - https://www.drupal.org/project/condition_query 
        - https://www.drupal.org/project/request_data_conditions
   - **Caching issues on simplytest.me and local ddev testing.**  

Long term, my hope is that https://www.drupal.org/project/drupal/issues/2916876 will work with modules like https://www.drupal.org/project/request_parameter_condition when it is released. 

...

Tested a bit with both drupal core 10.x visibility settings on simplytest.me and layout builder context locally, and came up with pretty similar conclusions.

I struggled to get LBC to really work, and even when it did, ran into the same caching issues. With a little work on the caching stuff, it seems like it'd be really well-suited for something like role-based customizations on an intranet site, where there can be a small handful of contexts defined for all pages on a site. For one-off campaign focuses, or something like that, it could get unwieldy fast. I think we could make it work, but that it'd take a lot of overhead to basically make it work like the D10 block visibility work is aiming to do.

The D10 core visibility settings look great. I don't think they'll cover what we're looking for out of the box, but it looks pretty easy to tie into and use, where we could build a lightweight, homegrown option if the community doesn't end up providing a full-fledged solution by then .

If there were a push to have something like this available in the nearer term, I also think we could build a homegrown version of this functionality with the aim of swapping it out for core in the future; with the downfalls of creating work intended to be dropped in the future (and a few potential roadblocks in getting our custom stuff visible in third-party block settings everywhere it's needed, as opposed to the ease of using core settings).

</details>

**Comment by makurj** [comment]
Feedback from Tippie: we don't want this feature built now, and will wait for D10 core functionality.
